### PR TITLE
logs of apps endpoint

### LINKED ISSUE #40: Cannot see logs of apps
Seems, that we need to implement [apps](https://github.com/lf-edge/eve/blob/d7206d73688044c7e829e9be73bcb4f5301a3442/pkg/pillar/cmd/logmanager/logmanager.go#L909) endpoint to view logs of apps.

### PR DESCRIPTION
**Pull request by giggsoff**
Implementation of `/apps/instances/id/{uuid}/logs` endpoint

Fix #40 
Signed-off-by: Petr Fedchenkov <giggsoff@gmail.com>

### CODE REVIEW

**Comment by deitch** [comment]
I will review. I would like to see #44 go in first, as it is a bug fix impacting serviceability.

File: pkg/driver/device_manager.go [hunk 1]
```diff
@@ -34,10 +34,12 @@ type DeviceManager interface {
 	MaxMetricSize() int
 	// MaxRequestsSize return the default maximum request logs size in bytes for this device manager
 	MaxRequestsSize() int
+	// MaxAppLogsSize return the default maximum app logs size in bytes for this device manager
+	MaxAppLogsSize() int
 	// Database safe-to-print (without credentials) path to database
 	Database() string
 	// Init initialize the datastore. If the given database URL is invalid for this type of manager, return false. Return error for actual failures
-	Init(string, int, int, int, int) (bool, error)
+	Init(string, int, int, int, int, int) (bool, error)
```
Comment (deitch) [inline pkg/driver/device_manager.go:42]: This is fine for now, but we will have to start making this `Init()` saner (in a future PR), maybe passing it a `struct`

File: pkg/driver/device_manager.go [hunk 2]
```diff
@@ -69,6 +71,8 @@ type DeviceManager interface {
 	WriteInfo(*info.ZInfoMsg) error
 	// WriteLogs write a LogBundle message
 	WriteLogs(*logs.LogBundle) error
+	// WriteAppInstanceLogs write a AppInstanceLogBundle message for instance UUID
+	WriteAppInstanceLogs(uuid.UUID, uuid.UUID, *logs.AppInstanceLogBundle) error
```
Comment (deitch) [inline pkg/driver/device_manager.go:75]: * do we need to pass the instance UUID as well? Is that, at least, not available from the `AppInstanceLogBundle`?
* can you add to the comment why we need to pass the app UUID (as the app log message does not indicate it)?
* if we need to pass the instance UUID, please put named parameters in the `interface`, so consumers can know which one is first

File: pkg/driver/file/device_manager_file.go [hunk 3]
```diff
@@ -510,6 +522,7 @@ func (d *DeviceManager) initDevice(u uuid.UUID) error {
 			dir:     path.Join(devicePath, requestsDir),
 			maxSize: int64(d.maxRequestsSize),
 		},
+		AppLogs: map[uuid.UUID]common.BigData{},
```
Comment (deitch) [inline pkg/driver/file/device_manager_file.go:525]: Why is this different? Why is it not a `ManagedFile` as well?

File: pkg/driver/memory/device_manager_memory.go [hunk 4]
```diff
@@ -252,6 +262,7 @@ func (d *DeviceManager) DeviceRegister(cert, onboard *x509.Certificate, serial s
 		Metrics: &ByteSlice{
 			maxSize: d.maxMetricSize,
 		},
+		AppLogs: map[uuid.UUID]common.BigData{},
```
Comment (deitch) [inline pkg/driver/memory/device_manager_memory.go:265]: Same question here. And where does the member of the `AppLogs` get instantiated?

File: pkg/driver/redis/device_manager_redis.go [hunk 5]
```diff
@@ -433,6 +440,7 @@ func (d *DeviceManager) DeviceRegister(cert, onboard *x509.Certificate, serial s
 			name:   deviceRequestsStream + unew.String(),
 			client: d.client,
 		},
+		AppLogs: map[uuid.UUID]common.BigData{},
```
Comment (deitch) [inline pkg/driver/redis/device_manager_redis.go:443]: Same here.

**Review by deitch — CHANGES_REQUESTED** [review]

File: pkg/driver/device_manager.go (hunk 2 above)
Comment (giggsoff) [reply pkg/driver/device_manager.go:75]: No uuid inside https://github.com/lf-edge/eve/blob/0811608f0a67121028c55c76f6c767d5311b32a6/api/go/logs/log.pb.go#L228

File: pkg/driver/file/device_manager_file.go [hunk 6]
```diff
@@ -680,6 +693,39 @@ func (d *DeviceManager) WriteLogs(m *logs.LogBundle) error {
...
+	if m == nil {
+		return fmt.Errorf("invalid nil message")
+	}
+	// get the uuid
+	// check that the device actually exists
+	if !d.deviceExists(deviceID) {
+		return fmt.Errorf("unregistered device UUID: %s", deviceID)
+	}
+	if !d.appExists(deviceID, instanceID) {
+		d.devices[deviceID].AppLogs[instanceID] = &ManagedFile{
```
Comment (giggsoff) [inline pkg/driver/file/device_manager_file.go:720]: I instantiate AppLogs here. And inside the same functions for the another managers.

Comment (deitch) [reply pkg/driver/file/device_manager_file.go:720]: OK, I see it now. Makes sense. Thanks for clarifying.

File: pkg/driver/device_manager.go (hunk 2 above)
Comment (deitch) [reply pkg/driver/device_manager.go:75]: OK, in that case, let's make the interface be explicit about the named args, so we know which is which, and make the comment more explicit.

File: pkg/driver/device_manager.go [hunk 7]
```diff
@@ -34,10 +34,12 @@ type DeviceManager interface {
 	MaxMetricSize() int
 	// MaxRequestsSize return the default maximum request logs size in bytes for this device manager
 	MaxRequestsSize() int
+	// MaxAppLogsSize return the default maximum app logs size in bytes for this device manager
+	MaxAppLogsSize() int
 	// Database safe-to-print (without credentials) path to database
 	Database() string
 	// Init initialize the datastore. If the given database URL is invalid for this type of manager, return false. Return error for actual failures
-	Init(string, int, int, int, int) (bool, error)
+	Init(string, *common.MaxSizes) (bool, error)
```
Comment (deitch) [inline pkg/driver/device_manager.go:42]: Ah, you already got ahead and put it in a struct? Thanks, this is much better.

Comment (deitch) [reply pkg/driver/device_manager.go:42]: I think you should pass this by value `common.MaxSizes` and not by reference `*common.MaxSizes`. If you pass it by value, it implies it is changeable (which is shouldn't be). More importantly, you will need to check for `== nil` on every implementation of `Init()`, whereas if you pass it by value, you don't need to do that.

**Review by deitch — CHANGES_REQUESTED** [review]
Thanks for adding the cleaner `Init()`. One change around passing by value vs by reference, and we will be good to go.

**Review by deitch — APPROVED** [review]

**Pull request merged**